internal/fs/inode/dir.go

```diff
@@ -706,16 +707,25 @@ func (d *dirInode) readObjects(
 	for _, p := range listing.CollapsedRuns {
 		pathBase := path.Base(p)
 		dirName := NewDirName(d.Name(), pathBase)
-		if c, ok := cores[dirName]; ok && c.Type() == metadata.ExplicitDirType {
-			continue
-		}
+		if !d.isBucketHierarchical() {
```

Comment:
imo hns vs non hns check should be moved outside collapsed run, otherwise for every element in collpsed this is being checked